Fix type annotation of assert functions

bernt-matthias opened the pull request:
asserts: n, delta, min, max need to be str since the allow extensions: (K|M|G|T|P|E)i

solved by automatically converting bools, integers, and suffixed integers before calling the assertion functions. 

follow up on https://github.com/galaxyproject/galaxy/pull/12528

I guess this should wait for the refactoring of the tests that is currently ongoing in https://github.com/galaxyproject/galaxy/pull/13186

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

bernt-matthias committed:
asserts: n, delta, min, max need to be str

since the allow extensions: (K|M|G|T|P|E)i

follow up on https://github.com/galaxyproject/galaxy/pull/12528

mvdbeek commented on lib/galaxy/tool_util/verify/asserts/archive.py:
```diff
@@ -49,7 +49,7 @@ def _list_from_zip(bytes, path):
 
 
-def assert_has_archive_member(output_bytes, path, verify_assertions_function, children, all="false", n: int = None, delta: int = 0, min: int = None, max: int = None, negate: bool = False):
+def assert_has_archive_member(output_bytes, path, verify_assertions_function, children, all="false", n: str = None, delta: str = 0, min: str = None, max: str = None, negate: bool = False):
```
If they can be int as well this should be a Union, and if the default is None I think you need typing.Optional
```suggestion
def assert_has_archive_member(output_bytes, path, verify_assertions_function, children, all="false", n: Optional[Union[str, int]] = None, delta: Union[str, int] = 0, min: Optional[Union[str, int]] = None, max: Optional[Union[str, int]] = None, negate: bool = False):
```

bernt-matthias committed:
Allow int and str for assertion arguments

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

bernt-matthias committed:
be strict with annotated types

instead of allowing Union[int,str] we only allow int
and autoconvert before calling.

bernt-matthias committed:
use typing_extensions

for python 3.7 get_args, get_origin are not available

bernt-matthias committed:
fix some typing problems

bernt-matthias committed:
remove some linting for assertions

with the added typing -- in particular Optional[int] (ie Union[int, None]) --
the tests in the assertions would have gone much more complicated

since https://github.com/galaxyproject/galaxy/pull/12528 they are also covered
by the xsd and therefor not really needed

mvdbeek commented on lib/galaxy/tool_util/verify/asserts/__init__.py:
```diff
@@ -44,10 +51,20 @@ def verify_assertion(data, assertion_description):
+            if get_origin(assert_function_annotations[attribute]) is not None:
+                types = get_args(assert_function_annotations[attribute])
+            if issubclass(int, types):
+                args[attribute] = parse_bytesize(value)
```
If someone specified `123M` for an integer value they'll get `123000000` as the value, or am I misunderstanding this ? Maybe you could narrow this down by creating a ByteCount type or similar ?

bernt-matthias replied:
Do you mean that we should not allow `123Mi` which would be `132070244352 = 123*1024^3`

mvdbeek commented on lib/galaxy/tool_util/verify/asserts/archive.py:
```diff
@@ -49,11 +50,19 @@ def _list_from_zip(bytes, path):
+def assert_has_archive_member(output_bytes: bytes,
+                              path: str,
+                              verify_assertions_function: Callable,
+                              children: List[etree.Element],
```
You sure that's the only type that can be passed here ? I think this should also work with yaml test definitions

bernt-matthias replied:
Good point. Seems that the parser makes it a List[Dict] anyway:

https://github.com/galaxyproject/galaxy/blob/32a41700f3026aa24b151a3fd7e7205ee3fd5fee/lib/galaxy/tool_util/parser/xml.py#L741
https://github.com/galaxyproject/galaxy/blob/32a41700f3026aa24b151a3fd7e7205ee3fd5fee/lib/galaxy/tool_util/parser/yaml.py#L282

Will change it.

mvdbeek commented on lib/galaxy/tool_util/verify/asserts/xml.py:
```diff
@@ -19,53 +20,89 @@ def assert_is_valid_xml(output):
+def assert_element_text(output: str,
+                        path: str,
+                        verify_assertions_function: Callable,
+                        children: List[etree.Element],
```
Again I'm not sure this isn't too tight.

mvdbeek commented on lib/galaxy/tool_util/verify/asserts/xml.py:
```diff
@@ -19,53 +20,89 @@ def assert_is_valid_xml(output):
+def assert_xml_element(output: str,
+                       path: str,
+                       verify_assertions_function: Optional[Callable] = None,
+                       children: Optional[List[etree.Element]] = None,
```
Same here

bernt-matthias committed:
set correct type for children parameters

bernt-matthias committed:
fix test
